diff --git a/ext/bigdecimal/bigdecimal.c b/ext/bigdecimal/bigdecimal.c
index e0b7c01..fe71e5e 100644
--- a/ext/bigdecimal/bigdecimal.c
+++ b/ext/bigdecimal/bigdecimal.c
@@ -2510,6 +2510,7 @@ BigDecimal_new(int argc, VALUE *argv)
 	return GetVpValue(iniValue, 1);

       case T_FLOAT:
+	rb_warn("initializing BigDecimal with Float.");
 	if (mf > DBL_DIG+1) {
 	    rb_raise(rb_eArgError, "precision too large.");
 	}
